packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts

```diff
@@ -0,0 +1,310 @@
+    await CloudSDK(request, response, {
+      sitecoreEdgeUrl: this.config.cdpConfig.sitecoreEdgeUrl,
+      sitecoreEdgeContextId: this.config.cdpConfig.sitecoreEdgeContextId,
+      siteName: this.config.sitecoreSiteName,
+      cookieDomain: hostname,
+      enableServerCookie: true,
+    })
+      .addPersonalize({ enablePersonalizeCookie: true })
+      .initialize();
+  }
+
+  protected getVariantIds = async (
+    req: IncomingMessage,
+    language: string,
+    pathname: string
+  ): Promise<string[]> => {
+    // const startTimestamp = Date.now();
+    const timeout = this.config.cdpConfig.timeout;
+
+    // Get personalization info from Experience Edge
+    const personalizeInfo = await this.personalizeService.getPersonalizeInfo(
+      pathname,
+      language,
+      this.config.sitecoreSiteName
+    );
+    if (!personalizeInfo) {
+      // Likely an invalid route / language
+      debug.personalize('skipped (personalize info not found)');
+      return [];
+    }
+
+    if (personalizeInfo.variantIds.length === 0) {
+      debug.personalize('skipped (no personalization configured)');
+      return [];
+    }
+
+    const params = this.getExperienceParams(req);
+    const executions = this.getPersonalizeExecutions(personalizeInfo, language);
+    const identifiedVariantIds: string[] = [];
+    try {
+      await Promise.all(
+        executions.map((execution) =>
+          this.personalize(
+            {
+              friendlyId: execution.friendlyId,
+              variantIds: execution.variantIds,
+              params,
+              language,
+              timeout,
+            },
+            req
+          ).then((personalization) => {
+            const variantId = personalization.variantId;
+            if (variantId) {
+              if (!execution.variantIds.includes(variantId)) {
+                debug.personalize('invalid variant %s', variantId);
+              } else {
+                identifiedVariantIds.push(variantId);
+              }
+            }
+          })
+        )
+      );
+    } catch (e) {
+      debug.personalize('skipped, error %o', e);
+    }
+
+    if (identifiedVariantIds.length === 0) {
+      debug.personalize('skipped (no variant(s) identified)');
+      return [];
+    }
+    return identifiedVariantIds;
+  };
+
+  protected personalizeLayout(layoutData: LayoutServiceData, variantIds: string[]) {
+    if (!layoutData.sitecore?.route) {
+      debug.personalize('skipped (layout is empty)');
+      return layoutData;
+    }
+    const personalizeData = getGroomedVariantIds(variantIds);
+    const personalizedPlaceholders = personalizeLayout(
+      layoutData,
+      personalizeData.variantId,
+      personalizeData.componentVariantIds
+    );
+    layoutData.sitecore.route.placeholders = personalizedPlaceholders;
+    return layoutData;
+  }
+
+  protected getLanguage(layoutData: LayoutServiceData): string {
+    return layoutData.sitecore?.context?.language || 'en';
```

Comment:
I think we need to add another check for "defaultLanguage"
XMC Proxy currently accepts "defaultLanguage" as a part of the "serverBundle", however in this case here it's ignored and our default value is prioritized